src/x/server/server_test.go

```diff
@@ -117,6 +125,137 @@ func TestServe(t *testing.T) {
+	s, h, numAdded, numRemoved := testServer(testListenAddress, TLSPermissive, false)
+
+	var (
+		numClients  = 9
+		expectedRes []string
+	)
+
+	err := s.ListenAndServe()
+	require.NoError(t, err)
+	listenAddr := s.listener.Addr().String()
+
+	for i := 0; i < numClients; i++ {
+		var conn net.Conn
+		var err error
+		if i%2 == 0 {
+			conn, err = net.Dial("tcp", listenAddr)
+		} else {
+			conn, err = tls.Dial("tcp", listenAddr, &tls.Config{InsecureSkipVerify: true})
+
+		}
+		require.NoError(t, err)
+
+		msg := fmt.Sprintf("msg%d", i)
+		expectedRes = append(expectedRes, msg)
+
+		_, err = conn.Write([]byte(msg))
+		require.NoError(t, err)
+	}
+
+	for h.called() < numClients {
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	require.False(t, h.isClosed())
+
+	s.Close()
+
+	require.True(t, h.isClosed())
+	require.Equal(t, int32(numClients), atomic.LoadInt32(numAdded))
+	require.Equal(t, int32(numClients), atomic.LoadInt32(numRemoved))
+	require.Equal(t, numClients, h.called())
+	require.Equal(t, expectedRes, h.res())
+}
+
+func TestTLSEnforcedServerListenAndClose(t *testing.T) {
+	s, h, numAdded, numRemoved := testServer(testListenAddress, TLSEnforced, false)
+
+	var (
+		numClients  = 10
+		expectedRes []string
+	)
+
+	err := s.ListenAndServe()
+	require.NoError(t, err)
+	listenAddr := s.listener.Addr().String()
+
+	for i := 0; i < numClients; i++ {
+		var conn net.Conn
+		var err error
+		msg := fmt.Sprintf("msg%d", i)
+
+		if i%2 == 0 {
+			conn, err = net.Dial("tcp", listenAddr)
+		} else {
+			conn, err = tls.Dial("tcp", listenAddr, &tls.Config{InsecureSkipVerify: true})
+			expectedRes = append(expectedRes, msg)
+		}
+		require.NoError(t, err)
+
+		_, err = conn.Write([]byte(msg))
+		require.NoError(t, err)
+	}
+
+	for h.called() < numClients/2 {
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	require.False(t, h.isClosed())
+
+	s.Close()
+
+	require.True(t, h.isClosed())
+	require.Equal(t, int32(numClients)/2, atomic.LoadInt32(numAdded))
+	require.Equal(t, int32(numClients)/2, atomic.LoadInt32(numRemoved))
+	require.Equal(t, numClients/2, h.called())
+	require.Equal(t, expectedRes, h.res())
+}
+
+func TestMutualTLSServerListenAndClose(t *testing.T) {
+	s, h, numAdded, numRemoved := testServer(testListenAddress, TLSEnforced, true)
+
+	var (
+		numClients  = 9
+		expectedRes []string
+	)
+
+	err := s.ListenAndServe()
+	require.NoError(t, err)
+	listenAddr := s.listener.Addr().String()
+	cert, err := tls.LoadX509KeyPair("./testdata/client.crt", "./testdata/client.key")
+	require.NoError(t, err)
+
+	for i := 0; i < numClients; i++ {
+		var conn net.Conn
+		var err error
+		conn, err = tls.Dial("tcp", listenAddr, &tls.Config{InsecureSkipVerify: true, Certificates: []tls.Certificate{cert}})
+		require.NoError(t, err)
+
+		msg := fmt.Sprintf("msg%d", i)
+		expectedRes = append(expectedRes, msg)
+
+		_, err = conn.Write([]byte(msg))
+		require.NoError(t, err)
+	}
+
+	for h.called() < numClients {
+		time.Sleep(100 * time.Millisecond)
```

Comment:
Set a timeout on this--don't sleep forever if the test fails. You can use `require.Eventually` for this purpose.

Reply:
We use the old version of testify, which doesn't have the Eventually function. Added limit here.

Reply:
Ah too bad, that's good to know. Can you upgrade testify in a separate PR? Should (presumably) be easy.

Reply:
Ah too bad, that's good to know. Can you upgrade testify in a separate PR? Should be easy to do.